debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java

```diff
@@ -434,6 +426,55 @@ public void addDataCollectionNamesToSnapshot(P partition, List<String> dataColle
         }
     }
 
+    @Override
+    @SuppressWarnings("unchecked")
+    public void stopSnapshot(P partition, List<String> dataCollectionIds, OffsetContext offsetContext) {
+        context = (IncrementalSnapshotContext<T>) offsetContext.getIncrementalSnapshotContext();
+        if (context.snapshotRunning()) {
+            if (dataCollectionIds == null || dataCollectionIds.isEmpty()) {
+                LOGGER.info("Stopping incremental snapshot.");
+                try {
+                    // This must be called prior to closeWindow to ensure the correct state is set
+                    // to prevent chunk reads from triggering additional open/close events.
+                    context.stopSnapshot();
+
+                    // Clear the state
+                    window.clear();
+                    closeWindow(partition, context.currentChunkId(), offsetContext);
+
+                    progressListener.snapshotAborted(partition);
+                }
+                catch (InterruptedException e) {
+                    LOGGER.warn("Failed to stop snapshot successfully.", e);
+                }
+            }
+            else {
+                final List<String> expandedDataCollectionIds = expandDataCollectionIds(dataCollectionIds);
+                LOGGER.info("Removing '{}' collections from incremental snapshot", expandedDataCollectionIds);
+                for (String dataCollectionId : expandedDataCollectionIds) {
+                    final TableId collectionId = TableId.parse(dataCollectionId);
+                    if (currentTable != null && currentTable.id().equals(collectionId)) {
+                        window.clear();
+                        LOGGER.info("Removed '{}' from incremental snapshot collection list.", collectionId);
+                        tableScanCompleted(partition);
+                        nextDataCollection(partition);
```

Comment:
@Naros This will theoretically trigger the snapshot of the next table which can be afterward cancelled.

I wonder if the algortihm should work this way

1) Remove all tables for the list of snaphotted tables save the currently snapshotted
2) Stop the snapshot of the current table if the stopis requested
3) Stop the snapshot completely if no more tables are to be snapshotted

Reply:
That does seem more deterministic. I'll make the adjustment.